bootstore/src/schemes/v0/request_manager.rs

```diff
@@ -0,0 +1,401 @@
+        );
+
+        self.envelopes.push(Envelope {
+            to,
+            msg: Msg::Req(Request {
+                id: request_id,
+                type_: RequestType::Learn,
+            }),
+        });
+
+        request_id
+    }
+
+    // Track a new request
+    fn new_request(
+        &mut self,
+        expiry: Instant,
+        request: TrackableRequest,
+    ) -> Uuid {
+        let id = Uuid::new_v4();
+        self.requests.insert(id, request);
+        self.expiry_to_id.insert(expiry, id);
+        id
+    }
+
+    // Send a `GetShare` request to all connected peers
+    fn broadcast_get_share(
+        &mut self,
+        request_id: Uuid,
+        rack_uuid: Uuid,
+        connected_peers: &BTreeSet<Baseboard>,
+    ) {
+        let iter = connected_peers.iter().cloned().map(|to| Envelope {
+            to,
+            msg: Request {
+                id: request_id,
+                type_: RequestType::GetShare { rack_uuid },
+            }
+            .into(),
+        });
+        self.envelopes.extend(iter);
+    }
+
+    /// Is there an outstanding `LearnSent` request
+    pub fn has_learn_sent_req(&self) -> bool {
+        self.requests.values().any(|req| {
+            if let TrackableRequest::LearnSent { .. } = req {
+                true
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Is there an outstanding `InitRack` request
+    pub fn has_init_rack_req(&self) -> bool {
+        self.requests.values().any(|req| {
+            if let TrackableRequest::InitRack { .. } = req {
+                true
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Return any expired requests mapped to their request id
+    ///
+    /// This is typically called during `tick` callbacks.
+    pub fn expired(
+        &mut self,
+        now: Instant,
+    ) -> BTreeMap<Uuid, TrackableRequest> {
+        let mut expired = BTreeMap::new();
+        while let Some((expiry, request_id)) = self.expiry_to_id.pop_last() {
+            if expiry < now {
+                expired.insert(
+                    request_id,
+                    self.requests.remove(&request_id).unwrap(),
+                );
+            } else {
+                // Put the last request back. We are done.
+                self.expiry_to_id.insert(expiry, request_id);
+                break;
+            }
+        }
+        expired
+    }
+
+    /// Return `Some(true)` if initialization completed, `Some(false) if it did not
+    ///
+    /// Return `None` if there is no active rack initialization ongoing
+    pub fn on_init_ack(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+    ) -> Option<bool> {
+        if let Some(TrackableRequest::InitRack { acks, .. }) =
+            self.requests.get_mut(&request_id)
+        {
+            acks.received.insert(from);
+            if acks.received == acks.expected {
```

Comment:
If someone that wasn't invited (in `acks.expected`) sends us an ack, we'd get permanently stuck here.  Not sure if that's worth worrying about, though.

Reply:
This is *definitely* worth worrying about along with your [comment below](https://github.com/oxidecomputer/omicron/pull/3385#discussion_r1256549979). The reason is that in combination with that bug below, we could contact a node that is deliberately in the rack, but not supposed to be part of the trust quorum - because we are using it for something else. Think Dogfood.

Reply:
Fixed in [774ac90](https://github.com/oxidecomputer/omicron/pull/3385/commits/774ac90417f756fdbf8d141532b013f716cb2e43)